File: python/easy/minimum_absolute_difference_in_bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.ret = sys.maxint
        self.pre = -1
        
        def dfs_inorder(root):
            if not root:
                return
            dfs_inorder(root.left)
            if self.pre != -1:
                self.ret = min(self.ret, root.val-self.pre)
            self.pre = root.val
            dfs_inorder(root.right)
        
        dfs_inorder(root)
        return self.ret

Conclustion:
Python does not have reference concept; thus, we use class variable as reference to keep track of the recursive.

File: cpp/easy/minimum_absolute_difference_in_bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        if(root == NULL)
            return 0;
        int pre = -1; //because the node value is non-negative
        int ret = INT_MAX;
        dfs_inorder(root, pre, ret);
        return ret;
    }
    
    void dfs_inorder(TreeNode* root, int& pre, int& ret)
    {
        if(root == NULL)
            return;
        dfs_inorder(root->left, pre, ret);
        if(pre != -1)
            ret = min(ret, root->val-pre);
        pre = root->val; //get the pre value before go to the right subtree(the bst is always right side larger than root and                //left side)
        dfs_inorder(root->right, pre, ret);
    }
};


Conclusion:
when we perform inorder travel, we can got a sorted number list; in this case, we just need to perform the inorder travel and use
a variable to store the pre-value of the node, then caculate the min-distance of list.

This methond is a pattern! great algorithm issue
